runtime/ts/plan/planificator.ts

```diff
@@ -0,0 +1,145 @@
+
+export class Planificator {
+  static async create(arc: Arc, {userid, protocol}) {
+    const store = await Planificator._initStore(arc, {userid, protocol, arcKey: null});
+    const planificator = new Planificator(arc, userid, store);
+    planificator.requestPlanning();
+    return planificator;
+  }
+
+  arc: Arc;
+  userid: string;
+  consumer: PlanConsumer;
+  producer: PlanProducer;
+  search: string|null = null;
+
+  // In <0.6 shell, this is needed to backward compatibility, in order to (1)
+  // (1) trigger replanning with a local producer and (2) notify shell of the
+  // last activated plan, to allow serialization.
+  // TODO(mmandlis): Is this really needed in the >0.6 shell?
+  arcCallback: ({}) => void;
+  lastActivatedPlan: Recipe|null;
+
+  constructor(arc: Arc, userid: string, store: StorageProviderBase) {
+    this.arc = arc;
+    this.userid = userid;
+    this.producer = new PlanProducer(arc, store);
+    this.consumer = new PlanConsumer(arc, store);
+
+    this.lastActivatedPlan = null;
+    this.arcCallback = this._onPlanInstantiated.bind(this);
+    this.arc.registerInstantiatePlanCallback(this.arcCallback);
+  }
+
+  async requestPlanning(options = {}) {
+    await this.producer.producePlans(options);
+  }
+
+  async loadPlans() {
+    return this.consumer.loadPlans();
+  }
+
+  setSearch(search) {
+    search = search ? search.toLowerCase().trim() : null;
+    search = (search !== '') ? search : null;
+    if (this.search !== search) {
+      this.search = search;
+      const showAll = this.search === '*';
+      const filter = showAll ? null : this.search;
+      this.consumer.setSuggestFilter(showAll, filter);
+    }
+  }
+
+  registerPlansChangedCallback(callback) {
+    this.consumer.registerPlansChangedCallback(callback);
+  }
+
+  registerSuggestChangedCallback(callback) {
+    this.consumer.registerSuggestChangedCallback(callback);
+  }
+
+  dispose() {
+    this.arc.unregisterInstantiatePlanCallback(this.arcCallback);
+    this.consumer.dispose();
+  }
+
+  getLastActivatedPlan() {
+    return {plan: this.lastActivatedPlan};
+  }
+
+  private _onPlanInstantiated(plan) {
+    this.lastActivatedPlan = plan;
+    this.requestPlanning();
+  }
+
+  private static async _initStore(arc, {userid, protocol, arcKey}) {
+    assert(userid, 'Missing user id.');
+    let storage = arc.storageProviderFactory._storageForKey(arc.storageKey);
+    const storageKey = storage.parseStringAsKey(arc.storageKey);
+    if (protocol) {
+      storageKey.protocol = protocol;
+    }
+    storageKey.location = storageKey.location
+        .replace(/\/arcs\/([a-zA-Z0-9_\-]+)$/, `/users/${userid}/suggestions/${arcKey || '$1'}`);
+    const storageKeyStr = storageKey.toString();
+    storage = arc.storageProviderFactory._storageForKey(storageKeyStr);
+    const schema = new Schema({names: ['Suggestions'], fields: {current: 'Object'}});
+    const type = Type.newEntity(schema);
+
+    // TODO: unify initialization of suggestions storage.
```

Comment:
Interesting ... why not use `createStore` instead of parsing storageKey manually?